crates/typst/src/layout/grid/layout.rs

```diff
@@ -598,36 +692,61 @@ impl<'a> GridLayouter<'a> {
                 }
 
                 // Render vertical lines.
-                for offset in points(self.rcols.iter().copied()) {
-                    let target = Point::with_y(frame.height() + thickness);
+                let auto_vlines =
+                    points(self.rcols.iter().copied()).enumerate().flat_map(|(x, dx)| {
+                        // We want each vline to span the entire table (start
+                        // at y = 0, end after all rows).
+                        // We use 'split_vline' to split the vline such that it
+                        // is not drawn above colspans.
+                        split_vline(self.grid, rows, x, 0, self.grid.rows.len())
+                            .into_iter()
+                            .zip(std::iter::repeat(dx))
+                    });
+                for ((dy, length), dx) in auto_vlines {
```

Comment:
Here I was considering that, in the future, we'll have custom lines which would probably have their own `dx` not generated through `points` (or at least not directly), so it'd make sense to have one loop handling vlines generically. However, one could argue that Part 4's design is not really finished yet, so it might not make much sense to think about that right now. I'll try to improve this part.